step functions (#18908)
DataDog/documentation: content/en/serverless/step_functions/troubleshooting.md

---
title: Troubleshooting Serverless Monitoring for AWS Step Functions
kind: documentation
---

## I cannot see any traces

### Verify that your Step Function is configured to send all logs
- Ensure that the `DD_TRACE_ENABLED` environment variable is set to `true`.
- In your AWS console, open your Step Function's logging tab. Ensure that _Log level_ is set to `ALL`, and that _Include execution data_ is selected.
- Ensure that the CloudWatch log group (also found on the logging tab) has a subscription filter to the Datadog Lambda Forwarder in the same region.

### Verify that logs are forwarded successfully to Datadog
- Check the Datadog Lambda Forwarder for error messages. Ensure that you have correctly set your API key and Datadog site.
- Enable `DEBUG` logs on the Datadog Lambda Forwarder by setting the environment variable `DD_LOG_LEVEL` to `debug`.

### Verify that logs are searchable on Live Search

> In Datadog, go to [*Logs*][1] > [*Live Tail*][2]. Search for `source:stepfunction`. You may need to trigger the state machine a few times.

Reviewer: For consistency, it's bolded in the section below. Also, since the end goal is to send user to Live Tail, I recommend cutting out the link for the Logs page.
```suggestion
In Datadog, go to [**Logs > Live Tail**][2]. Search for `source:stepfunction`. You may need to trigger the state machine a few times.
```

#### Search historic logs
To enable searching historic logs, add a temporary index to the forwarded logs. In Datadog, go to **Logs** > **Configuration** and then open the [**Indexes**][4] tab. Click the **New Index** button in the upper right.

Choose a name, set the index filter to `Source:stepfunction`, leave everything else with default values, and save.

{{< img src="serverless/step_functions/log_index.png" alt="New Log index" style="width:80%;" >}}

If your organization has an existing all-encompassing index with a low limit, place your new index at the top.

> **Note**: Log indexation is not a requirement for getting traces and may incur additional cost. If you are troubleshooting a specific issue, you may wish to temporarily turn on indexation, debug, and turn off indexation afterwards.

Reviewer: Indexing logs or Logs indexing.
```suggestion
**Note**: Indexing logs is not a requirement for getting traces and may incur additional cost. If you are troubleshooting a specific issue, you may wish to temporarily send logs to an index, debug, and delete the index afterwards. See [Indexes][6] for more information.
```

## Lambda traces are not merging with Step Function traces
- Verify that you can see both Lambda traces and Step Function traces in Datadog.
- Verify that you are using Python layer v75+ or Node.js layer v94+.
- In your AWS console, open your Step Function and ensure that your state machine has `"Payload.$": "States.JsonMerge($$, $, false)"` on the Lambda steps.
- Execute your Step Function once and verify that the `TaskScheduled` event log of the Lambda step has the payload containing data from the [Step Function context object][3].

#### Notes
Lambda steps that use the legacy Lambda API cannot be merged. If your Lambda step's definition is `"Resource": "<Lambda function ARN>"` instead of `"Resource": "arn:aws:states:::lambda:invoke"`, then your step is using the legacy Lambda API.

If your Lambda has the `DD_TRACE_EXTRACTOR` environment variable set, its traces cannot be merged.

[1]: https://app.datadoghq.com/logs
[2]: https://app.datadoghq.com/logs/livetail
[3]: https://docs.aws.amazon.com/step-functions/latest/dg/input-output-contextobject.html

> [4]: https://app.datadoghq.com/logs/pipelines/indexes

Reviewer: ```suggestion
[4]: https://app.datadoghq.com/logs/pipelines/indexes
[6]: /logs/log_configuration/indexes/
```

---
title: Troubleshooting Serverless Monitoring for AWS Step Functions
kind: documentation
---

## I cannot see any traces

### Verify that your Step Function is configured to send all logs

> - Ensure that the `DD_TRACE_ENABLED` environment variable is set to `true`.

Reviewer 2: Set `DD_TRACE_ENABLED` tag to `true` on the state machine.
